src/server/engine_shard_set.cc

```diff
@@ -56,6 +71,67 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
   return *this;
 }
 
+// This function checks 3 things:
+// 1. Don't try memory fragmentation if we don't use "enough" memory (control by
+// mem_defrag_threshold flag)
+// 2. That we had change in memory usage - to prevent endless loop - out of scope for now
+// 3. in case the above is OK, make sure that we have a "gap" between usage and commited memory
+// (control by commit_use_threshold flag)
+bool EngineShard::DefragTaskArgs::IsRequired() {
+  static const uint64_t mem_size = max_memory_limit;
+  static const float threshold_mem = mem_size / GetFlag(FLAGS_mem_defrag_threshold);
+  static const float commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
+
+  if (cursor > 0) {
+    return true;
+  }
+
+  uint64_t commited = GetMallocCurrentCommitted();
+  uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
+
+  // we want to make sure that we are not running this to many times - i.e.
+  // if there was no change to the memory, don't run this
+  if (threshold_mem < commited && mem_in_use != 0 &&
+      (float(mem_in_use * commit_use_threshold) < commited)) {
+    // we have way more commited then actual usage
+    return true;
+  }
+
+  return false;
+}
+
+// for now this does nothing
+bool EngineShard::DoDefrag() {
+  // TODO: impl!
+  // Placeholder for now!
+  return defrag_args_.cursor > 0;
+}
+
+void EngineShard::DefragTaskArgs::Init() {
+  cursor = 0u;
+}
+
+// the memory defragmentation task is as follow:
+//  1. Check if memory usage is high enough
+//  2. Check if diff between commited and used memory is high enough
+//  3. Check if we have memory changes (to ensure that we not running endlessly). - TODO
+//  4. if all the above pass -> run on the shard and try to defragmented memory by re-allocating
+//  values
+//     if the cursor for this is signal that we are not done, schedule the task to run at high
+//     priority otherwise lower the task priority so that it would not use the CPU when not required
+uint32_t EngineShard::DefragTask() {
+  const auto shard_id = db_slice().shard_id();
+  bool required_state = defrag_args_.IsRequired();
+  if (required_state) {
+    VLOG(1) << shard_id << ": need to run defrag memory cursor state: " << defrag_args_.cursor;
+    if (DoDefrag()) {  // we didn't finish the scan
+      return util::ProactorBase::kOnIdleMaxLevel;
```

Comment:
I had this conversation with Roman, there is an issue with this however, as it is hard to tell what is good enough. I have to say that I ran this with high load, and the idle task was simply "waited"

Reply:
Also note that this is not really high priority, just higher priority, these tasks are running when you have the "time" to run them - i.e. when the CPU for the given shard has a spare time.

Reply:
my 2 cents: I think we should submit something rough (i..e 2 levels are fine for starters) and then everyone can play with it and suggest improvements.

Reply:
I will commit a new commit today
This will include:
- the full impl end 2 end - the detection of the defrag and the running of the defrag at the compact object context
- Unit test at the mi_malloc context - allocating memory that checking whether we can detect defrag issues.
- unit test at the shard level to detect defrag issues
This I think would address you're suggestion